src/crimson/osd/main.cc

```diff
@@ -21,33 +21,59 @@ void usage(const char* prog) {
   generic_server_usage();
 }
 
+std::pair<std::vector<const char*>, std::vector<const char*>>
+partition_args(seastar::app_template& app, int argc, char* argv[])
+{
+  namespace bpo = boost::program_options;
+  // collect all options consumed by seastar::app_template
+  auto parsed = bpo::command_line_parser(argc, argv)
+    .options(app.get_options_description()).allow_unregistered().run();
+  auto unknown_args = bpo::collect_unrecognized(parsed.options,
+                                                bpo::include_positional);
+  std::vector<const char*> ceph_args, app_args;
+  for (auto arg = argv, end = argv + argc; arg != end; arg++) {
+    if (std::find(unknown_args.begin(), unknown_args.end(), *arg) !=
+        unknown_args.end()) {
+      ceph_args.push_back(*arg);
+    } else if (!std::strcmp(*arg, "-c")) {
+      // ceph_argparse_early_args() and
+      // seastar::smp::get_options_description() use "-c" for different options
+      if (arg + 1 != end) {
+        ceph_args.push_back(*arg++);
+        ceph_args.push_back(*arg);
+      }
+    } else {
+      app_args.push_back(*arg);
+    }
+  }
+  return {ceph_args, app_args};
+}
+
 int main(int argc, char* argv[])
 {
-  std::vector<const char*> args{argv + 1, argv + argc};
-  if (ceph_argparse_need_usage(args)) {
+  seastar::app_template app;
```

Comment:
app.add_options() is used to add app (crimson-osd)options, so --mkfs ,--key --osd-uuid will be get by crimson-osd when you use config=app.configuration(), --smp is not set in app.add_options, so they are transfer to seastar framework. I have tested them.

Reply:
see https://groups.google.com/forum/#!msg/seastar-dev/Q-dvMAawODE/wKLvbQO_DAAJ

Reply:
could you please take a look at the commit message?

also please note, `key` and `osd_uuid` are ceph config options, and they should be consumed by `conf.parse_argv()`, not by the boost program option parser in `seastar::app_template`.

Reply:
after I read  https://groups.google.com/forum/#!msg/seastar-dev/Q-dvMAawODE/wKLvbQO_DAAJ this discusstion, I think seastar use app.add_options and app.configuration() to withdraw ceph args, then we can use conf.parse_argv() to parse ceph args.  
see below description :Seastar parses the command line options (`argv[]`) when it is passed to `app_template::run()`, looking for its own standard options. Therefore, it is not recommended that the application tries to parse `argv[]` on its own because the application might not understand some of the standard Seastar options and not be able to correctly skip them.
+ 
+Rather, applications which want to have command-line options of their own should tell Seastar's command line parser of these additional application-specific options, and ask Seastar's command line parser to recognize them too. Seastar's command line parser is actually the Boost library's `boost::program_options`. An application adds its own option by using the `add_options()` and `add_positional_options()` methods on the `app_template` to define options, and later calling `configuration()` to retrieve the setting of these options.

Reply:
in practice, ceph allows user to pass whatever settings Ceph supports in command line, shall we add *all* of them using `add_options()`? i am afraid we shall not. `key` and `osd_uuid` are just the tip of the iceberg.